src/main/java/com/microsoft/sqlserver/jdbc/NTLMAuthentication.java

```diff
@@ -0,0 +1,905 @@
+     */
+    private byte[] hmacMD5(final byte[] key, final byte[] data) throws InvalidKeyException {
+        SecretKeySpec keySpec = new SecretKeySpec(key, "HmacMD5");
+        context.mac.init(keySpec);
+        return context.mac.doFinal(data);
+    }
+
+    /**
+     * Gets the MD4 hash of input string
+     *
+     * @param str
+     *        input string
+     * @return MD4 hash
+     */
+    private byte[] MD4(final byte[] str) {
+        MD4 md = new MD4();
+        md.reset();
+        md.update(str);
+        return md.digest();
+    }
+
+    /**
+     * Gets the unicode of string
+     *
+     * @param str
+     *        string to convert to unicode
+     * @return unicode of string
+     */
+    private byte[] unicode(final String str) {
+        return str.getBytes(java.nio.charset.StandardCharsets.UTF_16LE);
+    }
+
+    /**
+     * Concatenates 2 byte arrays
+     *
+     * @param arr1
+     * @param arr2
+     * @return concatenated array
+     */
+    private byte[] concat(final byte[] arr1, final byte[] arr2) {
+        if (null == arr1 || null == arr2) {
+            return null;
+        }
+
+        byte[] temp = new byte[arr1.length + arr2.length];
+        System.arraycopy(arr1, 0, temp, 0, arr1.length);
+        System.arraycopy(arr2, 0, temp, arr1.length, arr2.length);
+        return temp;
+    }
+
+    /**
+     * Generates a key from password to get NTLMv2 hash Section 3.3.2 NTLM v2 Authentication
+     *
+     * @return NT response key
+     * @throws InvalidKeyException
+     */
+    private byte[] ntowfv2() throws InvalidKeyException {
+        return hmacMD5(context.passwordHash,
+                (null != context.upperUserName) ? unicode(context.upperUserName + context.domainName)
+                                                : unicode(context.domainName));
+    }
+
+    /**
+     * Computes the NT response and keys from the Challenge message
+     *
+     * Section 3.3.2 NTLM v2 Authentication
+     *
+     * <pre>
+     * Set temp to ConcatenationOf(Responserversion, HiResponserversion, 
+     * Z(6), Time, ClientChallenge, Z(4), ServerName, Z(4)) 
+     * Set NTProofStr to HMAC_MD5(ResponseKeyNT,  
+     * ConcatenationOf(CHALLENGE_MESSAGE.ServerChallenge,temp)) 
+     * Set NtChallengeResponse to ConcatenationOf(NTProofStr, temp) 
+     * Set LmChallengeResponse to ConcatenationOf(HMAC_MD5(ResponseKeyLM,  
+     * ConcatenationOf(CHALLENGE_MESSAGE.ServerChallenge, ClientChallenge)), 
+     * ClientChallenge )
+     * </pre>
+     *
+     * @param responseKeyNT
+     *        NT response hash key
+     * @return computed response
+     * @throws InvalidKeyException
+     * @throws NoSuchAlgorithmException
+     */
+    private byte[] computeResponse(final byte[] responseKeyNT) throws InvalidKeyException, NoSuchAlgorithmException {
+        // get random client challenge nonce
+        byte[] clientNonce = new byte[NTLM_CLIENT_NONCE_LENGTH];
+        (new Random()).nextBytes(clientNonce);

```

Comment:
Do we want to change this to `ThreadLocalRandom.current().nextBytes(clientNonce);`?

>Normally to generate Random numbers, we either do Create an instance of java.util.Random OR Math.random() - which internally creates an instance of java.util.Random on first invocation. However in a concurrent applications usage of above leads to contention issues.
Random is thread safe for use by multiple threads. But if multiple threads use the same instance of Random, the same seed is shared by multiple threads. It leads to contention between multiple threads and so to performance degradation.
ThreadLocalRandom is solution to above problem. ThreadLocalRandom has a Random instance per thread and safeguards against contention.